File: Gas Station/solve.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    int canCompleteCircuit(vector<int> &gas, vector<int> &cost)
    {
        int leng = gas.size();
        int total_cost = 0;
        int total_reso = 0;
        int start = 0, current = 0;
        for (int i = 0; i < leng; i++)
        {
            total_cost += cost[i];
            total_reso += gas[i];
            current += (gas[i] - cost[i]);
            if (current < 0)
            {
                start = i + 1;
                current = 0;
            }
        }
        if (total_cost > total_reso)
            return -1;
        else
            return start;
    }
};

File: Gas Station/solve.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
var canCompleteCircuit = function (gas, cost) {
    const leng = gas.length;
    let total_gas_available = 0,
        total_burnt_gas = 0,
        current_resource = 0,
        start = 0;
    for (let i = 0; i < leng; i++) {
        total_burnt_gas += cost[i];
        total_gas_available += gas[i];
        current_resource += gas[i] - cost[i];
        if (current_resource < 0) {
            start = (i + 1) % leng;
            current_resource = 0;
        }
    }
    if (total_burnt_gas > total_gas_available) return -1;
    else return start;
};
